plugins/org.eclipse.glsp.api/src/org/eclipse/glsp/api/protocol/ClientSessionManager.java

```diff
@@ -0,0 +1,35 @@
+/********************************************************************************
+ * Copyright (c) 2020 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+package org.eclipse.glsp.api.protocol;
+
+import java.util.Optional;
+
+public interface ClientSessionManager {
+
+   boolean connectClient(GLSPClient client);
+
+   boolean createClientSession(GLSPClient glspClient, String clientId);
+
+   boolean disposeClientSession(String clientId);
+
+   boolean disconnectClient(GLSPClient client);
+
+   boolean addListener(ClientSessionListener listener);
+
+   boolean removeListener(ClientSessionListener listener);
+
+   Optional<GLSPClient> resolve(String clientId);
```

Comment:
Is this reliable? If I understand correctly, 1 GLSPClient corresponds to one "tab" in a browser, and 1 client ID corresponds to one "Widget" in a tab. So if we have 2 different GLSPClients, they may use the same clientId (Typically, xxx-diagram_0_widget). Since (AFAIU) the ClientSessionManager is shared across multiple modules, we can't assume that clientId will be unique ids (They are only unique for 1 GLSPClient / injection Module)

Reply:
Ah, great! Then it looks good to me :)